teps/0101-decouple-catalog-organization-and-reference.md

```diff
@@ -0,0 +1,233 @@
+The Tekton catalog is a collection of blessed Tekton resources that
+can be used with any system that supports the Tekton API. We also have
+the [hub](https://hub.tekton.dev) to be able to search for Tasks.
+
+This proposes to make the *Hub* and any other **official** mean of
+consuming Tekton resources from the community decoupled from where we
+author them. In a gist, this aims to make the repository
+`tektoncd/catalog` an implementation detail, and only support
+long-term, non-changing URI to refer to tasks (Those URIs could be
+`https://` endpoint such as the Hub or docker images, …).
+
+## Motivation
+
+As of today, users of the tekton task are directly refering to the
+GitHub URL to get Task from the catalog. This creates a hard coupling
+between the way we organize tasks inside Tekton and the way user
+consume it. This is forcing us into one possible catalog organization
+and prevent us to make drastic changes to this organization without
+breaking users.
+
+This enhancement aim to decouple the catalog organization from users
+consumption of those resources. This enhancement proposal is mainly
+touching the catalog and the hub components. This would also push the
+hub to be the main entrypoint for users to search, list and get tekton
+resources.
+
+### Goals
+
+- Allowing for the maintainers of `tektoncd/catalog` to change the
+  organization without affecting the consumption of the resources.
+- Define a set of standard, supported ways to get a tekton resource
+  (`Task`, `Pipeline`, …)
+
+### Non-Goals
+
+- Modify the current organization of the `tektoncd/catalog` repository.
+
+### Use Cases
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+## Requirements
+
+- A end-user should never refer / install task from a URL that points
+  directly to the github repository (like https://raw.githubusercontent.com/tektoncd/catalog/main/task/git-clone/0.5/git-clone.yaml).
+
+## Proposal
+
+This is a very very early draft, this will need to be better written and detailed, right now it’s just a “list” of things
+
+- User fetch tekton catalog resources throught the following “medium”
+  - oci image reference(s) : gcr.io/tekton-releases/catalog/upstream/task/{name}:{version}
+  - http endpoint : hub.tekton.dev/{catalog}/tasks/{name}/{version}
+- Two different catalog format envisionned
+  - legacy/file-based : current version of tektoncd/catalog
+    version are expressed as folder, resource organized as folder, …
+  - git-based: version are git versions (tag, …), meaning all resource from this are sharing the same version (can make sense for things that go together, for example a go pipeline and go tasks, for tests, builds, …)
```

Comment:
We discussed this a bit in the original design - I never particularly enjoyed having folder based versions, but using git versions also probably means that we need to tag the whole catalog to tag a single task, which I think it's the reason we didn't do it in the beginning. Do you have an new idea / proposal about that?

Reply:
yes, the idea here is, long-term, to split the catalog into several repositories. For example, grouping all the go task in the same repository (`tektoncd-catalog/golang`) so that we could use git-based versions instead, etc…

Doing the same parallel with the docker hub, all the "official" images (`docker.io/library/…`) lives in different repository for example. 

There is, however, a discussion to have on the hub on how to have way more flexibility to add catalogs, aka should users be able to have their repositories located in their org and register to the hub, etc..